ir/function.cpp

```diff
@@ -299,6 +374,112 @@ void Function::unroll(unsigned k) {
   LoopAnalysis la(*this);
   ofstream out("a.gv");
   la.printDot(out);
+
+  auto &nodes = la.getHeaderNodes();
+  auto &roots = la.getRoots();
+
+  BasicBlock &sink = getBB("#sink");
+
+  vector<pair<const BasicBlock*, bool>> worklist;
+
+  for (auto &root : roots) {
+    worklist.emplace_back(root, false);
+  }
+
+  while (!worklist.empty()) {
+    auto &[bb, flag]  = worklist.back();
+    if (flag) {
+      worklist.pop_back();
+      map<Value*, Value*> vmap;
+      map<const BasicBlock *, vector<const BasicBlock *>> bbmap;
+      vector<const BasicBlock *> cloned;
+      auto &bb_nodes = nodes[bb];
+
+      for (auto &&bb_ii : bb_nodes) {
+        bbmap[bb_ii].push_back(bb_ii);
+      }
+      for (unsigned unroll = 1 ; unroll <= k; ++unroll) {
+        for (auto &&bb_ii : bb_nodes) {
+          auto &bb_i = const_cast<BasicBlock&>(*bb_ii);
+          string twine = "_" + bb->getName() + "_" + to_string(unroll);
+          auto bb_cloned = &cloneBB(bb_i, twine, vmap);
+          bbmap[bb_ii].push_back(bb_cloned);
+          cloned.push_back(bb_cloned);
+        }
+      }
+
+      vector<const BasicBlock *> backedge_srcs;
+      set<const BasicBlock *> nodes_set(bb_nodes.begin(), bb_nodes.end());
+      // rewire targets
+      for (unsigned unroll = 0 ; unroll <= k; ++unroll) {
+        for (auto &&bb_ii : bb_nodes) {
+          auto curr = const_cast<BasicBlock *>(bbmap[bb_ii][unroll]);
+          map<const BasicBlock *, const BasicBlock *> replacement;
+          for (auto &tgt : curr->targets()) {
+            if (&tgt == bb || bb_ii != bb)
+              backedge_srcs.push_back(bb_ii);
+            // Filter out exit edges.
+            if (!nodes_set.count(&tgt))
+              continue;
+
+            const BasicBlock *to = nullptr;
+            // backedge
+            if (&tgt == bb) {
+              if (unroll == k)
+                to = &sink;
+              else
+                to = bbmap[&tgt][unroll + 1];
+            }
+            // edges inside loop
+            else {
+              to = bbmap[&tgt][unroll];
+            }
+            replacement[&tgt] = to;
```

Comment:
What about immediately calling `curr->replaceTargetWith(&tgt, to)` without using `replacement`?